numpy-stubs/__init__.pyi

```diff
@@ -388,22 +389,25 @@ class ndarray(_ArrayOrScalarCommon, Iterable, Sized, Container):
     def __contains__(self, key) -> bool: ...
     def __index__(self) -> int: ...
 
-class generic(_ArrayOrScalarCommon):
-    def __init__(self, value: Any = ...) -> None: ...
+class generic(_ArrayOrScalarCommon, metaclass=ABCMeta):
```

Comment:
I think a problem here is that `generic` is not actually an ABC; e.g.

```python
>>> inspect.isabstract(np.generic)
False
>>> np.generic.__abstractmethods__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: __abstractmethods__
```

Reply:
What I'm proposing here is to cheat a little bit for the sake of typing, considering that the  meta-class does (as far as mypy can tell) not affect the actual class anyway:
``` python
from abc import ABCMeta

class A: ...
class B(metaclass=ABCMeta): ...

A.__abstractmethods__
B.__abstractmethods__
```

```
>>> mypy test.py
test.py:6: error: "Type[A]" has no attribute "__abstractmethods__"
test.py:7: error: "Type[B]" has no attribute "__abstractmethods__"
Found 2 errors in 1 file (checked 1 source file)
```

Reply:
I think it does though; e.g.

```python
In [1]: import abc

In [2]: class B(metaclass=abc.ABCMeta):
   ...:     @abc.abstractmethod
   ...:     def foo(self): pass
   ...:

In [3]: B.__abstractmethods__
Out[3]: frozenset({'foo'})
```

OTOH I can't think of anything _better_ to do... will keep pondering.

Reply:
I think i figured out why mypy isn't picking up on the ``__abstractmethods__`` attribute. A closer look at the ``abc`` module in typeshed shows that it's rather incomplete ([ref](https://github.com/python/typeshed/blob/6fc882860b09842c4418c2d61e37407cf681e227/stdlib/3/abc.pyi#L8-L9)), only having the ``ABCMeta.register()`` method defined. 

I initially thought mypy's lack of ``__abstractmethods__`` recognition might have been a feature, but the relevant ``abc`` stub file just seems incomplete (and might thus be changed in the future). With this in mind I'm less confident in the use of ``ABCMeta``, even though, as far as I'm aware, it is the only way to explicitly disallow the initialization of a class.

Reply:
After a bit of testing I may have found a compromise: use the `@abstractmethod` decorator without the `ABCMeta` metaclass.
It does mean that mypy will warn about a lack of appropriate metaclass, but these errors can easily be silenced with `# type: ignore`. 

A (slimmed down) piece of example code:
``` python
class generic(_ArrayOrScalarCommon):
    @abstractmethod
    def __init__(self, *args: Any, **kwargs: Any) -> None: ...

class _real_generic(generic): ...  # type: ignore

class bool_(_real_generic):
    def __init__(self, value: object = ...) -> None: ...
```

Reply:
This seems like a reasonable compromise so long as we document it extensively (though it would be awkward if a third-party were subclassing the scalar hierarchy because then the ignores would leak into their code). A thing we should check on though is what other type checkers will do with this. We're pretty mypy heavy, but I don't want to use hacks that preclude people from using other type checkers.

It's also possible that it's better to allow the incorrect initialization for now and think about whether NumPy can be updated to make generic a true abc.

Reply:
I just did a few tests with a number of other type checkers ([pyright](https://github.com/Microsoft/pyright), [pyre](https://github.com/facebook/pyre-check) and [pytype](https://github.com/google/pytype)). With the sole exception of pyright they all seem to properly pick up the `@abstractmethod` decorator (just like mypy). Pyright, specifically, will remain silent unless the `ABCMeta` has been assigned as metaclass.

Used test code
----------------
``` python
from typing import Any
from abc import abstractmethod

class generic():
    @abstractmethod
    def __init__(self, *args: Any, **kwargs: Any) -> None: ...

class _real_generic(generic): ...  # type: ignore

class bool_(_real_generic):
    def __init__(self, value: object = ...) -> None: ...

a = generic()
b = _real_generic()
c = bool_()
```

Reply:
Awesome, thanks for checking that. Seems like this is a good solution then, just make sure to add comments about what's happening (and probably point to this discussion in the comment).